gematria/datasets/basic_block_utils_test.cc

```diff
@@ -0,0 +1,116 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gematria/datasets/basic_block_utils.h"
+
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "X86RegisterInfo.h"
+#include "gematria/llvm/asm_parser.h"
+#include "gematria/llvm/disassembler.h"
+#include "gematria/llvm/llvm_architecture_support.h"
+#include "gematria/testing/matchers.h"
+#include "gtest/gtest.h"
+
+using testing::UnorderedElementsAre;
+
+namespace gematria {
+namespace {
+class BasicBlockUtilsTest : public ::testing::Test {
+ private:
+  inline static std::unique_ptr<LlvmArchitectureSupport> LlvmArchSupport;
+
+ protected:
+  static void SetUpTestSuite() {
+    LlvmArchSupport = LlvmArchitectureSupport::X86_64();
+  }
+
+  std::vector<DisassembledInstruction> getInstructions(
+      std::string_view TextualAssembly) {
+    auto MCInstsOrError = gematria::ParseAsmCodeFromString(
+        LlvmArchSupport->target_machine(), TextualAssembly,
+        llvm::InlineAsm::AsmDialect::AD_ATT);
+    CHECK_OK(MCInstsOrError);
+
+    std::vector<DisassembledInstruction> Instructions;
+    Instructions.reserve(MCInstsOrError->size());
+
+    for (MCInst Instruction : *MCInstsOrError)
+      Instructions.push_back({0, "", "", Instruction});
+
+    return Instructions;
+  }
+
+  std::vector<unsigned> getUsedRegisters(std::string_view TextualAssembly) {
+    return BasicBlockUtils::getUsedRegisters(
+        getInstructions(TextualAssembly), LlvmArchSupport->mc_register_info(),
+        LlvmArchSupport->mc_instr_info());
+  }
+};
+
```

Comment:
A couple of other fun cases to test, maybe with some instructions before and after to make sure that the uses and defs are tracked correctly.

1. Lots of implicit uses and defs:
```asm
cpuid
```
2. `%ebx` is both a use and a def, implicit def of `%eflags`:
```asm
add %eax, %ebx
```
3. register aliasing is so much fun:
```asm
mov %al, 1      ; Write to a small part of RAX, rest is preserved.
add %eax, %ebx  ; As weird as it looks, RAX is still an input to the block.
```
4. register aliasing is even more fun:
```asm
mov %eax, 1     ; Cleans the upper bits of RAX (consistency FTW)
add %rax, %rbx  ; RAX is not an input of the block.
```
4. Addressing modes use registers too, let's make sure they are accounted for
```asm
add %rax, -16(%rbx, %rcx, 8)
```
6. Implicit use of `%eflags`, explicit output
```asm
seta %al
```
7. Implicit uses, implicit defs:
```asm
rep movsb
```

Reply:
Added most of these in (and fixed issues with register aliasing). I didn't add the last two as number six is just testing use of the flag register, which we're not accounting for, and `rep movsb` is kind of redundant with testing for `cpuid`/`rdtsc`.